src/audio/pcm_converter/pcm_converter_generic.c

```diff
@@ -25,22 +25,36 @@
 #include <stddef.h>
 #include <stdint.h>
 
+#define BYTES_TO_S16_SAMPLES	1
+#define BYTES_TO_S32_SAMPLES	2
+
 #if CONFIG_PCM_CONVERTER_FORMAT_S16LE && CONFIG_PCM_CONVERTER_FORMAT_S24LE
 
 static int pcm_convert_s16_to_s24(const struct audio_stream *source,
 				  uint32_t ioffset, struct audio_stream *sink,
 				  uint32_t ooffset, uint32_t samples)
 {
-	uint32_t buff_frag = 0;
-	int16_t *src;
-	int32_t *dst;
-	uint32_t i;
-
-	for (i = 0; i < samples; i++) {
-		src = audio_stream_read_frag_s16(source, buff_frag + ioffset);
-		dst = audio_stream_write_frag_s32(sink, buff_frag + ooffset);
-		*dst = *src << 8;
-		buff_frag++;
+	int16_t *src = source->r_ptr;
+	int32_t *dst = sink->w_ptr;
+	int nmax, i, n;
+	int processed = 0;
+
+	src = audio_stream_wrap(source, src + ioffset);
+	dst = audio_stream_wrap(sink, dst + ooffset);
+	while (processed < samples) {
+		n = samples - processed;
+		nmax = audio_stream_bytes_without_wrap(source, src) >> BYTES_TO_S16_SAMPLES;
+		n = MIN(n, nmax);
+		nmax = audio_stream_bytes_without_wrap(sink, dst) >> BYTES_TO_S32_SAMPLES;
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i++) {
+			*dst = *src << 8;
+			src++;
+			dst++;
+		}
+		processed += n;
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
```

Comment:
the `while` loop probably doesn't run more than once usually, right? Or just a couple of times at most. So, it's a rather minor optimisation, but these two assignments to `src` and `dst` on the last loop run are always redundant. You could do instead:
```
src += ioffset;
dst += ooffset;
for (processed = 0; processed < samples; processed += n) {
	src = audio_stream_wrap(source, src);
	dst = audio_stream_wrap(sink, dst);
...
}
```
and remove these two calls to `audio_stream_wrap()` - in all cases probably

Reply:
Thanks, I think this did good for code size since this repeats many times!